File: Semestre_2/Vision_Computadora/Entregas/Matlab/T2_13_Fotomontaje2.m
clear;
close all;
clc;

% Cartel
[file1, path1] = uigetfile({'*.png';'*.jpeg';'*.jpg'}, 'File Selector');
ruta1 = strcat(path1,file1);
M = im2double(imread(ruta1));

% Persona
[file2, path2] = uigetfile({'*.png';'*.jpeg';'*.jpg'}, 'File Selector');
ruta2 = strcat(path2,file2);
A = im2double(imread(ruta2));
A = histeq(A);              %Aplicacion de ecualizado
A =1.0*(A/1.0).^(1/1.8);    %Aplicacion de gama=1.8

% Fondo
[file3,path3] = uigetfile({'*.png';'*.jpeg';'*.jpg'}, 'File Selector');
ruta3=strcat(path3,file3);
F = im2double(imread(ruta3));

[fx fy fz] = size(F);
M=imresize(M,[fx fy]);
A=imresize(A,[fx fy]);

% Calcular la diferenca D = abs(M-A)
D = abs(M - A);

% Umbralizar U=D(D,x)
if length(size(D))==3 % Si es RGB
    Dgris=rgb2gray(D);
else
    Dgris = D;
end
umbral=0.225;
U = Dgris >= umbral; 

% Calcular R = (F AND NOT U) OR (A AND U)
NOT_U = 1 - U;
parte1 = F.*NOT_U;
parte2 = A.*U;
R = parte1 + parte2;

%4. Visualización del fotomontaje
figure;
subplot(3,2,1);imshow(M);title('M');
subplot(3,2,2);imshow(A);title('A');
subplot(3,2,3);imshow(Dgris);title('D');
subplot(3,2,4);imshow(U);title('U');
subplot(3,2,5);imshow(F);title('F');
subplot(3,2,6);imshow(R);title('R');

File: Semestre_2/Vision_Computadora/Entregas/Python/T2_13_Fotomontaje2.py
import matplotlib.pyplot as plt
import numpy as np
from skimage.io import imread
from skimage.color import rgb2gray
from tkinter import Tk, filedialog
from skimage.exposure import equalize_hist
from skimage.transform import resize

Tk().withdraw()

# Cartel
file_path1 = filedialog.askopenfilename(
    title='File Selector - Cartel',
    filetypes=[("Images", "*.png;*.jpeg;*.jpg")]
)
M = imread(file_path1).astype(float) / 255.0

# Persona
file_path2 = filedialog.askopenfilename(
    title='File Selector - Persona',
    filetypes=[("Images", "*.png;*.jpeg;*.jpg")]
)
A = imread(file_path2).astype(float) / 255.0

if A.ndim == 3 and A.shape[-1] == 4:
    A = A[..., :3]

if A.ndim == 3:
    A = np.stack([equalize_hist(A[..., c]) for c in range(A.shape[-1])], axis=-1)
else:
    A = equalize_hist(A)

A = A ** (1/1.8)

# Fondo
file_path3 = filedialog.askopenfilename(
    title='File Selector - Fondo',
    filetypes=[("Images", "*.png;*.jpeg;*.jpg")]
)
F = imread(file_path3).astype(float) / 255.0

if F.ndim == 3 and F.shape[-1] == 4:
    F = F[..., :3]
if M.ndim == 3 and M.shape[-1] == 4:
    M = M[..., :3]

fx, fy = F.shape[:2]
M = resize(M, (fx, fy), preserve_range=True)
A = resize(A, (fx, fy), preserve_range=True)

D = np.abs(M - A)

if D.ndim == 3:
    Dgris = rgb2gray(D)
else:
    Dgris = D

umbral = 0.225
U = Dgris >= umbral

U_f = U.astype(float)
NOT_U = 1.0 - U_f

if F.ndim == 3:
    U_f3 = U_f[..., np.newaxis]
    NOT_U3 = NOT_U[..., np.newaxis]
else:
    U_f3 = U_f
    NOT_U3 = NOT_U

parte1 = F * NOT_U3
parte2 = A * U_f3
R = parte1 + parte2

plt.figure()
plt.subplot(3,2,1); plt.imshow(M); plt.title('M'); plt.axis('off')
plt.subplot(3,2,2); plt.imshow(A); plt.title('A'); plt.axis('off')
plt.subplot(3,2,3); plt.imshow(Dgris, cmap='gray'); plt.title('D'); plt.axis('off')
plt.subplot(3,2,4); plt.imshow(U, cmap='gray'); plt.title('U'); plt.axis('off')
plt.subplot(3,2,5); plt.imshow(F); plt.title('F'); plt.axis('off')
plt.subplot(3,2,6); plt.imshow(R); plt.title('R'); plt.axis('off')
plt.show()
